---
title: "MR of DNAm on FEV1"
output: html_notebook
---

```{r}
library(readxl)
library(TwoSampleMR)
library(tidyverse)
library(ieugwasr)
library(here)
```

Following an EWAS of FEV1, there is interest in investigating the causal relationship that underlies the associations. Barring technical artefacts, EWAS signals can plausibly represent associations due to DNAm being on the causal pathway to lung function, lung function status influencing DNAm levels, or confounders of DNAm and lung function. Mendelian randomisation in conjunction with colocalisation analysis can attempt to separate these mechanisms through use of known mQTLs as genetic instruments. 

DNAm sites have a cis region which often harbour a single cis-acting mQTL, and occasionally DNAm sites also have known trans-acting mQTLs. Some DNAm sites have multiple independent cis effects however these are difficult to determine reliably due to reliance on accurately capturing sample LD using an external reference - so for these analysis we will only use independent instruments.

1. Gather and organise data
2. Perform basic MR analysis of each DNAm site on FEV1
3. Test if FEV1(EWAS) related mQTLs are enriched for low p-values in FEV1 GWAS
4. Test if DNAm - FEV1 MR analyses support causal influence of DNAm on FEV1
5. Colocalisation to optionally follow up


## 1. Gather and organise data

We will perform MR analysis of DNAm on FEV1. The FEV1 GWAS used will be:

```{r}
gwasinfo("ebi-a-GCST007432") %>% str()
```

The mQTL results will be obtained from the GoDMC summary data which is available at https://github.com/MRCIEU/godmc_phase2_analysis.

The list of DNAm sites to read Read in list of CpGs

```{r}
ewas_result <- read_excel(here("EWAS_PFTs_FDRsignCpGs_450K_MKL.xlsx"), col_names=TRUE, sheet=1, skip=1)
fev1_cpg <- ewas_result$`CpG Probe`
str(ewas_result)
```

Get GoDMC mQTLs

```{r}
load(url("https://github.com/MRCIEU/godmc_phase2_analysis/raw/master/results/16/16_clumped.rdata"))
clumped <- ungroup(clumped)
```

Pull in rsids

```{r}
load("snp_1kg.rdata")
str(snp_1kg)
m <- match(clumped$snp, snp_1kg$snp)
clumped$rsid <- NA
table(clumped$snp == snp_1kg$snp[m])
clumped$rsid[!is.na(m)] <- na.exclude(snp_1kg$V2[m])
```


Format clumped data for MR analysis

```{r}
exposure_dat <- clumped %>%
  ungroup() %>%
  filter(cpg %in% fev1_cpg) %>%
  {TwoSampleMR::format_data(
    dat=.,
    snp_col="rsid", 
    phenotype_col="cpg",
    beta_col="Effect",
    se_col="StdErr",
    effect_allele_col="Allele1",
    other_allele_col="Allele2",
    pval_col="pval",
    samplesize_col="TotalSampleSize",
    chr_col="snpchr",
    pos_col="snppos"
  )}
str(exposure_dat)
```

Of `r nrow(ewas_result)` DNAm sites, `r length(unique(exposure_dat$exposure))` have at least one mQTL detected in whole blood from the GoDMC meta analysis results. This is the distribution of number of independent mQTLs per DNAm site

```{r}
exposure_dat %>%
  group_by(exposure) %>%
  summarise(n_mqtl=n()) %>%
  group_by(n_mqtl) %>%
  summarise(count=n())
```

Extract mQTL variants from FEV1 GWAS

```{r}
outcome_dat <- extract_outcome_data(unique(exposure_dat$SNP), "ebi-a-GCST007432")
str(outcome_dat)
```

Harmonise exposure and outcome data

```{r}
dat <- harmonise_data(exposure_dat, outcome_dat, action=1) %>% suppressMessages()
```

## 2. Perform basic MR analysis of each DNAm site on FEV1

```{r}
res <- mr(dat, method_list=c("mr_ivw", "mr_wald_ratio")) %>% suppressMessages
save(res, dat, ewas_result, file="output.rdata")
```

Quick look at the MR results:

```{r}
res <- inner_join(
  res, 
  clumped %>% select(cpg, cpgchr, cpgpos) %>% filter(!duplicated(cpg)), by=c("exposure"="cpg")) %>% 
  mutate(cpgchr = as.numeric(gsub("chr", "", cpgchr)), fdr=p.adjust(pval, "fdr"))
```

```{r}
ggplot(res %>% filter(method %in% c("Inverse variance weighted", "Wald ratio")), aes(x=cpgpos, y=b)) +
  geom_point(aes(colour=fdr < 0.05, shape=method)) +
  facet_grid(. ~ cpgchr, scale="free_x", space="free_x") +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.position="bottom") +
  labs(x="DNAm site location", y="MR estimate")
ggsave("mr_miami.pdf", width=10, height=5)
```

There are some associations here that are stronger than expected by chance e.g. what fraction of the MR results have an FDR < 0.05? (5% of these are expected to be false positives)

```{r}
prop.table(table(res$fdr < 0.05))
```

And what fraction of MR results have a p-value < 0.05? (expect 0.05 under the null hypothesis of no associations)

```{r}
prop.table(table(res$pval < 0.05))
```

## 3. Test if FEV1(EWAS) related mQTLs are enriched for low p-values in FEV1 GWAS

Is this number of significant mQTLs in FEV1 more than expected by chance (given that FEV1 is polygenic and that mQTLs are enriched for high LD, intermediate MAF etc)? Compare the test statistics of the FEV1 mQTLs against the rest of the mQTLs in GoDMC, adjusting for SNP variance and LD scores.

```{r}
mqtl_snps <- unique(clumped$rsid) %>% na.exclude()
length(mqtl_snps)
```

Extract all mQTLs from FEV1

```{r}
mqtl_fev1 <- extract_outcome_data(mqtl_snps, "ebi-a-GCST007432", proxies=FALSE)
```

What fraction of all GoDMC mQTLs associate with FEV1?

```{r}
prop.table(table(mqtl_fev1$pval.outcome < 0.05))
```

Compares to FEV1 mQTLs

```{r}
prop.table(table(mqtl_fev1$pval.outcome[mqtl_fev1$SNP %in% dat$SNP] < 0.05))
```

Hypothesis - FEV1 EWAS related mQTLs are more likely to relate to FEV1 GWAS than all other known mQTLs

Obtain allele frequency and LD score info for the mQTLs

```{r}
afl2 <- afl2_rsid(mqtl_snps)
```

Perform the analysis of

$$
FEV1_{GWAS} \sim mQTL_{EWAS} + LDscore + Var(SNP)
$$

i.e. if we denote an $mQTL_{EWAS}$ to be for `DNAm sites associated with FEV1 in EWAS` vs `all other DNAm sites`, then does that predict lower GWAS p-values for FEV1?

```{r}
afl2$fev1 <- as.numeric(afl2$rsid %in% exposure_dat$SNP)
enr_dat <- inner_join(
  afl2,
  mqtl_fev1 %>% select(rsid=SNP, beta=beta.outcome, se=se.outcome) %>% mutate(fval=beta^2/se^2),
  by="rsid"
)
enr_dat$snpvar <- (1-enr_dat$AF.EUR) * enr_dat$AF.EUR
model1 <- summary(lm(fval ~ fev1 + L2.EUR + snpvar, enr_dat))
model1
```

This indicates that mQTLs for FEV1 EWAS hits associate more strongly than all other mQTLs - i.e. the EWAS is picking up DNAm sites that capture some sort of FEV1 regulatory function.

Also try binarising the FEV1 GWAS test statistic (i.e. is the test statistic for FEV1 GWAS p < 0.05)

```{r}
enr_dat$sig <- enr_dat$fval > qf(0.05, 1, 300000, low=F)
model2 <- summary(glm(sig ~ fev1 + L2.EUR + snpvar, enr_dat, family="binomial"))
model2
```

Similar result - an mQTL for an FEV1 associated CpG is `r model2$coef[2,1] %>% exp() %>% round(., 2)` times more likely to associate with FEV1 than any other mQTL.

## 4. Test if DNAm - FEV1 MR analyses support causal influence of DNAm on FEV1

The MR results indicate that the FEV1 EWAS is capturing DNAm sites that are genetically influenced by genetic factors that also influence FEV1. One model is that there is a causal path (vertical pleiotropy) 

```
G -> DNAm -> FEV1
```

Another model is horizontal pleiotropy

```
DNAm <- G -> FEV1
```

Under the vertical pleiotropy model, we would expect that the effect estimates of DNAm on FEV1 are consistent when estimated using different instruments for DNAm. Many of the DNAm sites do have multiple independent instruments, so we can try to perform this analysis.

We will formulate the analysis by hypothesising that the IVW association for `DNAm -> FEV1` is substantially stronger when combined across all instruments, than if we use any one instrument alone

```{r}
res2 <- mr_singlesnp(dat, all_method="mr_ivw")

mcpg <- filter(res2, grepl("All", SNP))$exposure %>% unique

comparison <- res2 %>% 
  filter(exposure %in% mcpg) %>%
  filter(!is.na(p)) %>%
  group_by(exposure) %>%
  do({
    x <- .
    t1 <- x %>% 
      filter(!grepl("All", SNP)) %>%
      {tibble(exposure=.$exposure[1], minpval=min(.$p), n=nrow(.))}
    t2 <- x %>%
      filter(grepl("All", SNP)) %>%
      {tibble(exposure=.$exposure[1], ivwpval=.$p[1])}
    inner_join(t1, t2, by="exposure")
  })

head(comparison)
table(comparison$minpval > comparison$ivwpval)
comparison[comparison$minpval > comparison$ivwpval, ]
```

We find `r sum(comparison$minpval > comparison$ivwpval)` instances out of `r nrow(comparison)` in which the IVW association is stronger than the strongest Wald ratio. However, looking at the results these are very modest differences. To test if this is more than we'd expect by chance, shuffle the SNPs across DNAm sites and re-estimate the joint effects - is `r sum(comparison$minpval > comparison$ivwpval)` more than we'd expect by chance?

```{r}
set.seed(12345)
nshuf <- 10
shufres <- rep(0, nshuf)
for(i in 1:nshuf)
{
  message(i)
  shuffle <- sample(1:nrow(dat))
  dat2 <- dat
  dat2$exposure <- dat2$exposure[shuffle]
  dat2$id.exposure <- dat2$id.exposure[shuffle]
  res3 <- mr_singlesnp(dat2, all_method="mr_ivw")
  mcpg <- filter(res3, grepl("All", SNP))$exposure %>% unique
  
  comparison_shuffle <- res3 %>% 
    filter(exposure %in% mcpg) %>%
    filter(!is.na(p)) %>%
    group_by(exposure) %>%
    do({
      x <- .
      t1 <- x %>% 
        filter(!grepl("All", SNP)) %>%
        {tibble(exposure=.$exposure[1], minpval=min(.$p), n=nrow(.))}
      t2 <- x %>%
        filter(grepl("All", SNP)) %>%
        {tibble(exposure=.$exposure[1], ivwpval=.$p[1])}
      inner_join(t1, t2, by="exposure")
    })
  shufres[i] <- sum(comparison_shuffle$minpval > comparison_shuffle$ivwpval) / nrow(comparison_shuffle)
}
```

```{r}
tibble(
  prop=c(shufres, sum(comparison$minpval > comparison$ivwpval) / nrow(comparison)), 
  what=c(rep("Random permutations", nshuf), "Real data")
) %>%
  arrange(prop) %>%
  mutate(ord=1:n()) %>%
  ggplot(., aes(x=ord, y=prop)) +
    geom_point(aes(colour=what))
```

Here we see that the rate of stronger associations due to combining across DNAm sites is no more than expected by chance under a model of no causality.


## 5. Colocalisation

The MR analysis doesn't necessarily support a causal role of DNAm on FEV1, however this analysis could be expanded by performing colocalisation of DNAm and FEV1 at the mQTLs that genetically associate with FEV1.

Determine list of DNAm-SNP sets to perform coloc on

```{r}
tocoloc <- res2 %>%
  filter(!grepl("All", SNP)) %>%
  mutate(fdr = p.adjust(p, "fdr")) %>%
  filter(fdr < 0.01)
save(tocoloc, file="tocoloc.rdata")
```

Obtain regional mQTL genetic effects from GoDMC API

```{r}
library(httr)
library(dplyr)
library(tidyr)

query <- list(
    cpgs = unique(tocoloc$exposure)[1:3]
)

o <- lapply(unique(tocoloc$exposure), function(cpg)
  {
    o <- POST("http://api.godmc.org.uk/v0.1/query", body = list(cpgs=cpg), encode = "json")
    content(o) %>% lapply(., as_tibble) %>% bind_rows
  })
```

Organise regional data

```{r}
cpgregions <- o %>% bind_rows() %>%
  separate(name, into=c("snpchr", "snppos", "snptype"), sep=":") %>%
  mutate(
    snpchr=as.numeric(gsub("chr", "", snpchr)),
    snppos=as.numeric(snppos)
  )
str(cpgregions)
```

Perform colocalisation for each DNAm-SNP set that associated with FEV1 GWAS

```{r, message=FALSE, echo=FALSE}
colocres <- lapply(1:nrow(tocoloc), function(i)
{
  CPG <- tocoloc$exposure[i]
  RSID <- tocoloc$SNP[i]
  a <- subset(cpgregions, cpg == CPG)
  CHR <- subset(a, rsid == RSID)$snpchr
  POS <- subset(a, rsid == RSID)$snppos
  region <- paste0(CHR, ":", POS-100000, "-", POS+100000)
  a <- subset(a, snpchr==CHR & snppos > (POS-100000) & snppos < (POS+100000))
  if(nrow(a) < 10)
  {
    return(NULL)
  }
  out <- suppressMessages(extract_outcome_data(a$rsid, "ebi-a-GCST007432"))
  aexp <- format_data(a, snp_col="rsid", beta_col="beta_a1", se_col="se", effect_allele_col = "a1", other_allele_col = "a2", samplesize_col = "samplesize", phenotype_col = "cpg", eaf_col = "freq_a1")
  colocdat <- suppressMessages(harmonise_data(aexp, out, action=1))
  if(nrow(colocdat) < 10)
  {
    return(NULL)
  }
  d1 <- list(beta=colocdat$beta.exposure, varbeta=colocdat$se.exposure^2, type="quant", N=colocdat$samplesize.exposure, sdY=1, snp=colocdat$SNP, pvalues=colocdat$pval.exposure)
  d2 <- list(beta=colocdat$beta.outcome, varbeta=colocdat$se.outcome^2, type="quant", N=colocdat$samplesize.outcome, sdY=1, snp=colocdat$SNP, pvalues=colocdat$pval.outcome)
  maf <- colocdat$eaf.exposure
  maf[maf > 0.5] <- 1 - maf[maf > 0.5]
  sink("/dev/null")
  cr <- suppressMessages(suppressWarnings(coloc.abf(d1, d2, MAF=maf)))
  sink()
  x <- tibble(rsid=RSID, cpg=CPG, chr=CHR, pos=POS, h4=cr$summary['PP.H4.abf'])
  return(x)
}) %>% bind_rows()
```

Distribution of P(H4)

```{r}
hist(colocres$h4)
```

How many colocalise at p(h4) > 0.8?

```{r}
sum(colocres$h4 > 0.8)
```

In regions of very long LD e.g. MHC region, colocalisation is difficult to interpret. This is because colocalisation aims to determine if two traits share a causal variant. However, if the region has extended LD then there will be a very large number of potential causal variants in high LD with each other, and it's possible that the causal variant for DNAm is different to the causal variant for FEV1 and this will not be detectable by coloc.

```{r}
table(colocres$h4 > 0.8, colocres$chr)
```

Here are the results that colocalise that are not on chromosome 6

```{r}
colocres %>%
  filter(chr !=6 & h4 > 0.8)
```

```{r}
save(colocres, file="colocres.rdata")
```

